nvflare/recipe/spec.py

```diff
@@ -37,31 +38,33 @@ def __init__(self, job: FedJob):
         """
         self.job = job
 
-    def add_input_filter_to_clients(self, filter: Filter, tasks=None):
+    def add_client_input_filter(self, filter: Filter, tasks=None, clients=Optional[List[str]]):
         """Add a filter to clients for incoming tasks from the server.
 
         Args:
             filter: the filter to be added
             tasks: tasks that the filter applies to
+            clients: client names to add
 
         Returns: None
 
         """
         self.job.to_clients(filter, filter_type=FilterType.TASK_DATA, tasks=tasks)
 
-    def add_output_filter_to_clients(self, filter: Filter, tasks=None):
+    def add_client_output_filter(self, filter: Filter, tasks=None, clients=Optional[List[str]]):
         """Add a filter to clients for outgoing result to server.
 
         Args:
             filter: the filter to be added
             tasks: tasks that the filter applies to
+            clients: client names to add
 
         Returns: None
 
         """
         self.job.to_clients(filter, filter_type=FilterType.TASK_RESULT, tasks=tasks)
```

Comment:
If clients is explicitly specified, you cannot call job.to_clients() any more - it only works to add to ALL clients.
Instead, you need to call job.to_client() for each client in the "clients" list.